### PR TITLE
[PROD-2276] Add new class to cached auth token to file (or redis with subclass)

### PR DESCRIPTION
**Pull request by romainissynced**
# Summary

- Changed the auth flow so that you can pass a CachedToken cls
- Made new class to cache token in file cache
- Made sure that the file cache works with the lowest py version that we support, py 3.7 

## Checklist

Before formally opening this PR, please adhere to the following standards:

- [x] Branch/PR names begin with the related Jira ticket id (ie PROD-31) for Jira integration
- [x] File names are lower_snake_case
- [x] Relevant unit tests have been added or not applicable
- [x] Relevant documentation has been added or not applicable
- [x] Mark yourself as the assignee (makes it easier to scan the PR list)

[Related Jira Ticket](https://synccomputing.atlassian.net/browse/PROD-2276)

Add any relevant testing examples or screenshots.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -50,6 +50,7 @@ dev = [
     "Sphinx==4.3.0",
     "respx==0.20.1",
     "deepdiff==6.3.0",
+    "python-dateutil==2.9",
```
Comment (romainissynced) [inline pyproject.toml:53]: this was already installed from botocore, pinning the version here to be safe

File: sync/clients/sync.py [hunk 2]
```diff
@@ -13,78 +14,110 @@
...
-        if self._access_token_expires_at_utc:
-            return datetime.now(
-                tz=timezone.utc
-            ) < self._access_token_expires_at_utc - timedelta(seconds=20)
-        return False
+
+class SyncAuth(httpx.Auth):
+    requires_response_body = True
+
+    def __init__(self, api_url: str, api_key: APIKey, cached_token_cls: Type[CachedToken] = FileCachedToken):
```
Comment (romainissynced) [inline sync/clients/sync.py:99]: The plan here is to subclass `CachedToken` from `sync-svc-library-tasks` to support redis caching, the file cache behavior should match what was there before.

File: sync/clients/sync.py [hunk 3]
```diff
@@ -94,11 +127,9 @@ def build_auth_request(self) -> httpx.Request:
     def update_access_token(self, response: httpx.Response):
         if response.status_code == httpx.codes.OK:
             auth = response.json()
-            self._access_token = auth["result"]["access_token"]
-            self._access_token_expires_at_utc = auth["result"]["expires_at_utc"]
-            self._set_cached_token(
-                self._access_token, self._access_token_expires_at_utc
-            )
+            access_token = auth["result"]["access_token"]
+            expires_at_utc = dateutil.parser.isoparse(auth["result"]["expires_at_utc"])
```
Comment (romainissynced) [inline sync/clients/sync.py:131]: Had to do this because py3.7 can't parse a datetime with timezone with builtin `fromisoformat`

Comment (shughes-uk) [reply sync/clients/sync.py:131]: boo!

**Review by shughes-uk — DISMISSED** [review]

File: pyproject.toml [hunk 4]
```diff
@@ -35,7 +35,8 @@ dependencies = [
     "azure-mgmt-compute==30.1.0",
     "azure-mgmt-resource==23.0.1",
     "azure-cli-core==2.50.0",
-    "platformdirs"
+    "platformdirs",
+    "python-dateutil==2.9",
```
Comment (shughes-uk) [inline pyproject.toml:39]: I suspect this pin will cause some pain fyi

Comment (singhals) [reply pyproject.toml:39]: Is this internal pain or for our customers?

Comment (shughes-uk) [reply pyproject.toml:39]: our customers, strict pins like this are somewhat painful.

Comment (singhals) [reply pyproject.toml:39]: Ya agreed - lets figure out a good plan after this pr gets out to address the pinned versions.

**Review by singhals — APPROVED** [review]

File: pyproject.toml (hunk 4 above)
Comment (romainissynced) [reply pyproject.toml:39]: Made a ticket to track that: https://synccomputing.atlassian.net/browse/PROD-2324

Can bring it up at parking lot, would pinning as <3 be safe as that's the next major version that might introduce breaks?

**Pull request merged**